# template-nf

<!-- This README.md is the single page user documentation for this pipeline. -->

The `bwa-meme` implementation of mapping and duplicate removal.

## Pipeline description

Performs mapping using the `bwa mem` algorithm which significant performance enhancements.

`bwa-meme` is based on `bwa-mem2` with further improvements by learned index.

## Input

### `--input`

A csv file of input fastq pairs (optionally gz compressed).

Format:

```csv
rgId,rgLb,rgPl,rgSm,rgPu,read1,read2
1,sampA_lib,ILLUMINA,sampA,R1_L1,testdata/1_1.fq.gz,testdata/1_2.fq.gz
2,sampA_lib,ILLUMINA,sampA,R1_L2,testdata/2_1.fq.gz,testdata/2_2.fq.gz
```

### `--genome_fasta`

Path to reference genome fasta, co located files expected:

- genome.dict (`samtools dict`)
- genome.fa.{0123,alt,amb,ann,pac}
- genome.fa.{bwt.2bit.64,pos_packed}
- genome.fa.suffixarray_uint64_L{1,2}_PARAMETERS

All `genome.fa.*` files generated by following `bwa-meme` [indexing][meme-index] and [training][meme-train].

## Output

### `--outdir`

Output folder for pipeline, expect:

- `execution_report_*.html`
- `execution_timeline_*.html`
- `execution_trace_*.txt`
- `logs/` split by `${task.process}/${task.index}`
- `marked.met` - samtools markdup metrics file (like picard)

When `--format bam` (default):

- `marked.bam`
- `marked.bam.bai`

When `--format cram`:

- `marked.cram`
- `marked.cram.crai`

## Usage

Simplest form of execution:

```
nextflow run main.nf --genome_fasta genome.fasta --input data.csv [Options]
```

Ensure `-profile` and `--executor` set as resquired for your platform.

## Options

Check the pipeline help section (`nextflow main.nf --help`) for all the updated options and their default values.

### `--input`

CSV file, header plus one line per fastq pair of input.

### `--genome_fasta`

Path to `genome.fasta` file with co-located files described above

### `--format`

Defaults to `bam`.  Options `bam` or `cram`.

### `--bp_in_batch`

Defaults to `100000000`, only used in testing to ensure full reprodicability (`-K` in `bwa`/`bwa-mem2`/`bwa-meme`).

### `--outdir`

Path to write final results to.

<!-- For Sphinx doc, This option will be auto rendered help() section from Nextflow main.nf in the doc build -->


<!------------------
Build of this doc in github handle by - .github/workflows/build-deploy-doc.yml

To build this doc locally follow these steps.

Needs to have installed - 
1. sphinx
2. sphinx-rtd-theme
3. nextflow

Supposing your currently in base directory of the pipeline -
```
cd docs && bash src/pre-build.sh
cp README.md src
cd src && make html 
```
index.html will be generated in `docs/src/build/html` folder
-->

<!-- refs -->
[meme-index]: https://github.com/kaist-ina/BWA-MEME/#build-index-of-the-reference-dna-sequence
[meme-train]: https://github.com/kaist-ina/BWA-MEME/#training-p-rmi